File: BilisPera/BilisPera/Product/Home/Controller/ProductHomeViewController.m
//
//  ProductHomeViewController.m
//  BilisPera
//
//  Created by BHJ on 2025/8/14.
//

#import "ProductHomeViewController.h"
#import "ProductHomeHeaderView.h"

NS_ASSUME_NONNULL_BEGIN

@interface ProductHomeViewController ()<UITableViewDelegate,UITableViewDataSource>

@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) ProductHomeViewModel *viewModel;
@property (nonatomic, strong) UIButton *nextButton;

@end

@implementation ProductHomeViewController

- (instancetype)initWith:(NSString *)productId
{
    self = [super init];
    if (self) {
        self.productId = productId;
    }
    return self;
}

- (void)viewDidLoad{
    [super viewDidLoad];
    [self configViewModel];
    [self configUI];
}

- (void)viewDidAppear:(BOOL)animated{
    [super viewDidAppear:animated];
    [self.tableView.mj_header beginRefreshing];
}

-(void)reloadData{
    kWeakSelf;
    [self.viewModel reloadDataWithProductId:self.productId completion:^{
        [weakSelf.tableView reloadData];
        [weakSelf.tableView.mj_header endRefreshing];
    }];
}

- (UITableView *)tableView{
    if (!_tableView) {
        _tableView = [[UITableView alloc] init];
        _tableView.backgroundColor = [UIColor clearColor];
        _tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
        _tableView.separatorColor = [UIColor clearColor];
        _tableView.estimatedRowHeight = 44.0;
        _tableView.estimatedSectionFooterHeight = 0;
        _tableView.estimatedSectionHeaderHeight = 0;
        _tableView.delegate = self;
        _tableView.dataSource = self;
    }
    return _tableView;
}

- (UIButton *)nextButton{
    if (!_nextButton) {
        _nextButton = [UIButton buttonWithType:UIButtonTypeCustom];
        _nextButton.backgroundColor = kColor_351E29;
        _nextButton.layer.cornerRadius = kRatio(22);
        _nextButton.layer.masksToBounds = YES;
        [_nextButton setTitle:@"Confirm" forState:UIControlStateNormal];
        [_nextButton setTitleColor:kWhiteColor forState:UIControlStateNormal];
        _nextButton.titleLabel.font = kFont(18);
        [_nextButton addTarget:self action:@selector(nextStep) forControlEvents:UIControlEventTouchUpInside];
    }
    return _nextButton;
}

-(void)configUI{
    self.navigationItem.title = @"Product details";
    [self.view addSubview:self.tableView];
    [self.tableView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.view).offset(kNavigationBarHeight);
        make.leading.trailing.equalTo(self.view);
        make.bottom.equalTo(self.view);
    }];
    
    [self.view addSubview:self.nextButton];
    [self.nextButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.leading.trailing.equalTo(self.view).inset(kRatio(16));
        make.bottom.equalTo(self.view).inset(kSafeAreaBottomHeight + kRatio(20));
        make.height.mas_equalTo(kRatio(44));
    }];
    kWeakSelf;
    self.tableView.mj_header = [MJRefreshHeader headerWithRefreshingBlock:^{
        [weakSelf reloadData];
    }];
}

-(void)configViewModel{
    self.viewModel = [[ProductHomeViewModel alloc] init];
    [self.tableView registerClass:[ProductHomeHeaderView class] forCellReuseIdentifier:[[ProductHomeHeaderView alloc] init].reuseId];
    [self.tableView registerClass:[ProductHomeStepCell class] forCellReuseIdentifier:[[ProductHomeStepCell alloc] init].reuseId];
}

// MARK: - UITableViewDelegate,UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return self.viewModel.dataSource.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    ProductSectionModel *setionModel = self.viewModel.dataSource[section];
    return setionModel.cellData.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    ProductSectionModel *setionModel = self.viewModel.dataSource[indexPath.section];
    BaseTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:setionModel.cellId forIndexPath:indexPath];
    [cell configData:setionModel.cellData[indexPath.row]];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
}


-(void)nextStep{
    [[ProductHandle shared] onPushNextStep:self.productId];
}
@end

NS_ASSUME_NONNULL_END

File: SkyLoan_phillipines/Product/ProductHomeViewController.swift
//
//  ProductHomeViewController.swift
//  SkyLoan_phillipines
//
//  Created by BHJ on 2025/5/22.
//

import UIKit
import MJRefresh

class ProductHomeViewController: AuthenticationBaseController {
    convenience init(productId: String,orderNum: String){
        self.init()
        viewModel.productId = productId
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        viewModel.eventDelegate = self
        Task{
           await LoginTool.shared.requestAddressList()
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        hiddenNavigationBar = true
        listView.mj_header?.beginRefreshing()
    }
    
    var viewModel = ProductViewModel()
    
    lazy var headerView: ProductHomeHeaderView = {
        let view = ProductHomeHeaderView(frame: .init(x: 0, y: 0, width: kScreenW, height: 120.ratio()),model: .init())
        view.isHidden = true
        return view
    }()
}

extension ProductHomeViewController{
    var hiddenNavigationBar: Bool{
        true
    }
    
    override func setupUI(){
        super.setupUI()
        navBar.title = LocalizationConstants.Product.nav_title
        listView.register(ProductListCell.self, forCellReuseIdentifier: ProductListCell.reusableId)
        listView.delegate = self
        listView.dataSource = self
        view.addSubview(headerView)
        headerView.snp.makeConstraints { make in
            make.top.equalTo(navBar.snp.bottom).offset(12.ratio())
            make.leading.trailing.equalToSuperview()
            make.height.equalTo(120.ratio())
        }
        listView.snp.remakeConstraints { make in
            make.top.equalTo(headerView.snp.bottom)
            make.leading.trailing.bottom.equalToSuperview()
        }
        nextButton.setAttributedTitle(LocalizationConstants.Product.bottomButtonTitle.addUnderline(), for: .normal)
        listView.mj_header = MJRefreshHeader(refreshingBlock: {
            [weak self] in
            self?.reloadData()
        })
        view.bringSubviewToFront(navBar)
    }
    
    @MainActor
    override func reloadData(){
        Task{
            await viewModel.reloadData()
            headerView.model = viewModel.headerModel
            headerView.isHidden = false
            listView.reloadData()
            await listView.mj_header?.endRefreshing()
            guard let url = viewModel.getUrl(),!url.isEmpty else {return}
            HJPrint(url)
        }
    }
}

extension ProductHomeViewController: UITableViewDelegate,UITableViewDataSource{
    
    func numberOfSections(in tableView: UITableView) -> Int {
        viewModel.dataSource.count
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let sectionModel = viewModel.dataSource[section]
        return sectionModel.cellDatas?.count ?? 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let sectionModel = viewModel.dataSource[indexPath.section]
        let cell = tableView.dequeueReusableCell(withIdentifier: sectionModel.cellType.reusableId)
        if let model = sectionModel.cellDatas?[indexPath.row]{
            cell?.data = model
        }
        return cell ?? UITableViewCell()
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard viewModel.itemList.count > indexPath.row else {return}
        let model = viewModel.itemList[indexPath.row]
        itemClick(model: model)
    }
}


